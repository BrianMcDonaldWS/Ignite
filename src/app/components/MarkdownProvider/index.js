import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';

import App from '../App';

const setMarkdown = (markdown = []) => {
  return markdown.reduce(
    (markdownMap, [pathToMarkdown, component, isIndex, firstLink]) => {
      markdownMap[pathToMarkdown] = component;

      if (isIndex) {
        markdownMap.indexFiles = {
          ...markdownMap.indexFiles,
          [pathToMarkdown]: firstLink
        };
      }

      return markdownMap;
    },
    {}
  );
};

const setPlugins = plugins => {
  return plugins.reduce(
    (pluginMap, [name, component, options]) =>
      Object.assign({}, pluginMap, {
        [name]: {
          name,
          component,
          options
        }
      }),
    {}
  );
};

class MarkdownProvider extends Component {
  static propTypes = {
    location: ReactRouterPropTypes.location.isRequired,
    markdown: PropTypes.array,
    plugins: PropTypes.array
  };

  static defaultProps = {
    markdown: [],
    plugins: []
  };

  constructor(props) {
    super(props);

    this.markdown = setMarkdown(props.markdown);
    this.plugins = setPlugins(props.plugins);
  }

  render() {
    return (
      <App
        markdown={this.markdown}
        location={this.props.location}
        plugins={this.plugins}
      />
    );
  }
}

export default MarkdownProvider;
