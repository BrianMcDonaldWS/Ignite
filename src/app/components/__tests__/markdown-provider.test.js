import React from 'react';

import MarkdownProvider from '../MarkdownProvider';
import renderToJson from './utils/render-to-json';

test('Has default route', () => {
  expect(renderToJson(<MarkdownProvider />)).toMatchSnapshot();
});

const markdown = [];

test('update markdown provided', () => {
  markdown.push(['/foo/bar', () => <h1> Simple Component </h1>]);
  expect(
    renderToJson(<MarkdownProvider markdown={markdown} />, { new: true })
  ).toMatchSnapshot();
});

test('sets first page and root from index file', () => {
  markdown.push([
    '/foo/bar',
    () => <h1> Simple Component </h1>,
    true,
    '/first/page'
  ]);
  expect(
    renderToJson(<MarkdownProvider markdown={markdown} />, { new: true })
  ).toMatchSnapshot();
});

const plugins = [];

test('uses plugins', () => {
  plugins.push(['test', () => <h1> Simple Component </h1>]);
  expect(
    renderToJson(<MarkdownProvider markdown={markdown} plugins={plugins} />, {
      new: true
    })
  ).toMatchSnapshot();
});
