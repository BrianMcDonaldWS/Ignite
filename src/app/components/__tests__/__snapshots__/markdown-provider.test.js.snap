// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Has default route 1`] = `
<App
  index={undefined}
  location={undefined}
  markdown={Object {}}
  plugins={Object {}}
/>
`;

exports[`sets first page and root from index file 1`] = `
<App
  index={undefined}
  location={undefined}
  markdown={
    Object {
      "/foo/bar": [Function],
      "indexFiles": Object {
        "/foo/bar": "/first/page",
      },
    }
  }
  plugins={Object {}}
/>
`;

exports[`update markdown provided 1`] = `
<App
  index={undefined}
  location={undefined}
  markdown={
    Object {
      "/foo/bar": [Function],
    }
  }
  plugins={Object {}}
/>
`;

exports[`uses plugins 1`] = `
<App
  index={undefined}
  location={undefined}
  markdown={
    Object {
      "/foo/bar": [Function],
      "indexFiles": Object {
        "/foo/bar": "/first/page",
      },
    }
  }
  plugins={
    Object {
      "test": Object {
        "component": [Function],
        "name": "test",
        "options": undefined,
      },
    }
  }
/>
`;
